Let MySQLConnector accept all UTF charsets and recommend utf8mb4

hoesler opened the pull request:
Allow and prefer utf8mb4 character set in mysql metadata storage.

Upgrade Instructions:
Use ALTER DATABASE and ALTER TABLE as described in https://dev.mysql.com/doc/refman/5.7/en/charset-unicode-conversion.html

Closes #5377

drcrallen commented on extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java:
```diff
@@ -163,17 +162,19 @@ protected int getStreamingFetchSize()
       throw new ISE(
-          "Database default character set is not UTF-8." + System.lineSeparator()
-          + "  Druid requires its MySQL database to be created using UTF-8 as default character set."
+          "Druid requires its MySQL database to be created with an UTF charset, found `%1$s`.%n" +
```
Can you remove newlines ? It confuses lots of logging systems.

hoesler replied:
Sure.

drcrallen commented on extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java:
```diff
@@ -163,17 +162,19 @@ protected int getStreamingFetchSize()
         .first();
 
-    if (!isUtf8) {
+    if (!databaseCharset.matches("utf.*")) {
```
Is the result ever cased differently? Do we need a to lowercase?

drcrallen replied:
Don't this also match a bunch of other things like utf16?

hoesler replied:
Don't you think lowercase is okay, if it has been until now? But yes, it should be `uft8.*`. I'll fix it.

drcrallen commented:
Cool, https://github.com/druid-io/druid/blob/druid-0.11.0/docs/content/development/extensions-core/mysql.md needs updated, and can you add "upgrade instructions" to the PR master comment for those who used the "old" instructions for utf8 mysql databases?

hoesler commented:
I already updated the mysql.md. Is anything missing there?

For the upgrade instructions, I added a link to the mysql docs, as it is not generic. See here: https://dba.stackexchange.com/questions/8239/how-to-easily-convert-utf8-tables-to-utf8mb4-in-mysql-5-5

hoesler commented:
Alternatively, I could assemble all required statements to migrate in the mysql.md.

drcrallen commented:
Cool! 

@hoesler would you mind filling out http://druid.io/community/cla.html please?

hoesler commented:
Sorry for the delay. Signed it!

gianm commented:
@drcrallen I'm assuming you are good due to the positivity of your last comment (https://github.com/druid-io/druid/pull/5411#issuecomment-368383317). Let us know if otherwise.
